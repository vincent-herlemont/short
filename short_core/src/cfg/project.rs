use std::path::PathBuf;

pub trait ProjectCfg {
    fn path(&self) -> PathBuf;
}
