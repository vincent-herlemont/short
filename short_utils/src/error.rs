//! short specific Error type
//! Allow to embedded multiple error type.
use super::result::Result;
use promptly::ReadlineError;
/// Inspiration from : https://github.com/brson/basic-http-server/blob/1ab052719a88e41822b2955d7d72bf161457d47c/src/main.rs#L468
use serde::export::fmt::Debug;
use serde_yaml;

use std::error::Error as StdError;
use std::fmt::Result as FmtResult;
use std::fmt::{Display, Formatter};
use std::io;
use std::path::StripPrefixError;

use std::string::FromUtf8Error;

pub type ExitStatusCode = Option<i32>;

// TODO: Add PartialEq to Error type.
#[derive(Debug)]
pub enum Error {
    Other(String),
    Wrap(String, Box<Error>),
    Io(io::Error),
    SerdeYaml(serde_yaml::Error),
    Which(which::Error),
    StripPrefixError(StripPrefixError),
    ExitStatus(ExitStatusCode),
    RustyLine(ReadlineError),
    StringUtf8(FromUtf8Error),
}

impl Error {
    /// Create new error
    /// # Example
    /// Usually use [`format!`] as string parameter for pass context information.
    ///
    /// ```
    /// use short_utils::error::Error;
    /// let url = "http://.....";
    /// Error::new(format!("fail to get url {}",url));
    /// ```
    pub fn new<S: AsRef<str>>(msg: S) -> Error {
        Error::Other(String::from(msg.as_ref()))
    }

    /// Create new box error
    pub fn new_box(msg: String) -> Box<Error> {
        Box::new(Error::new(msg))
    }

    pub fn wrap<S: AsRef<str>>(msg: S, err: Error) -> Error {
        Error::Wrap(String::from(msg.as_ref()), Box::new(err))
    }

    pub fn is<P>(&self, predicate: P) -> bool
    where
        P: Fn(&Error) -> bool,
    {
        match self {
            Error::Wrap(_, err) => err.is(predicate),
            err => predicate(err),
        }
    }
}

impl Display for Error {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        match self {
            Error::Other(msg) => write!(f, "{}", msg),
            Error::Wrap(msg, err) => write!(f, "{} : {}", msg, err),
            Error::Io(err) => write!(f, "{}", err),
            _ => write!(f, "{:?}", self),
        }
    }
}

impl StdError for Error {
    fn source(&self) -> Option<&(dyn StdError + 'static)> {
        use Error::*;

        match self {
            Io(e) => Some(e),
            SerdeYaml(e) => Some(e),
            Which(_) => None,
            Wrap(_, error) => Some(error),
            Other(_) => None,
            StripPrefixError(e) => Some(e),
            ExitStatus(_) => None,
            RustyLine(e) => Some(e),
            StringUtf8(e) => Some(e),
        }
    }
}

impl From<&'static str> for Error {
    fn from(message: &'static str) -> Error {
        Error::Other(String::from(message))
    }
}

impl From<String> for Error {
    fn from(message: String) -> Error {
        Error::Other(message)
    }
}

impl From<io::Error> for Error {
    fn from(error: io::Error) -> Error {
        Error::Io(error)
    }
}

impl From<serde_yaml::Error> for Error {
    fn from(error: serde_yaml::Error) -> Error {
        Error::SerdeYaml(error)
    }
}

impl From<which::Error> for Error {
    fn from(error: which::Error) -> Error {
        Error::Which(error)
    }
}

impl From<StripPrefixError> for Error {
    fn from(error: StripPrefixError) -> Error {
        Error::StripPrefixError(error)
    }
}

impl From<ReadlineError> for Error {
    fn from(error: ReadlineError) -> Error {
        Error::RustyLine(error)
    }
}

impl From<ExitStatusCode> for Error {
    fn from(exit_status: ExitStatusCode) -> Error {
        Error::ExitStatus(exit_status)
    }
}

impl From<FromUtf8Error> for Error {
    fn from(error: FromUtf8Error) -> Error {
        Error::StringUtf8(error)
    }
}

impl Error {
    pub fn exit_code_eq(&self, code: i32) -> Result<bool> {
        if let Error::ExitStatus(exit_status) = self {
            let cmd_code = exit_status.ok_or(Error::new("Process terminated by signal"))?;
            return Ok(code == cmd_code);
        }
        Err(Error::new("no exit status")).unwrap()
    }
}

#[cfg(test)]
mod tests {
    use super::Error;
    use crate::result::Result;
    use std::io;
    use std::io::ErrorKind;

    #[test]
    fn error_test() {
        fn create_io_error() -> Result<()> {
            Err(Error::from(io::Error::new(io::ErrorKind::Other, "ho no !")))
        }
        fn create_other_error() -> Result<()> {
            Err(Error::new(format!("ho no !")))
        }
        fn create_other_error_from() -> Result<()> {
            Err(Error::from("ho no !"))
        }

        let result = create_io_error();
        match result.err().unwrap() {
            Error::Io(_) => assert!(true),
            _ => assert!(false),
        };

        let result = create_other_error();
        match result.err().unwrap() {
            Error::Other(_) => assert!(true),
            _ => assert!(false),
        };

        let result = create_other_error_from();
        match result.err().unwrap() {
            Error::Other(_) => assert!(true),
            _ => assert!(false),
        };
    }

    #[test]
    fn nested_error() {
        let err = Error::from(io::Error::new(ErrorKind::NotFound, "not found"));
        let err = Error::wrap("wrap lvl 1", err);
        let err = Error::wrap("wrap lvl 2", err);
        let b = err.is(|err| matches!(err, Error::Io(_)));
        assert!(b);
        let b = err.is(|err| matches!(err, Error::Other(_)));
        assert!(!b);
    }
}
