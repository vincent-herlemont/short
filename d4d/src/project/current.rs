use serde::{Deserialize, Serialize};
use std::fs::{create_dir, File, OpenOptions};
use std::io::{BufReader, BufWriter};
use std::path::{Path, PathBuf};
use utils::error::Error;
use utils::result::Result;

#[allow(dead_code)]
const PROJECT_CONFIG_DIRECTORY: &'static str = ".d4d";
#[allow(dead_code)]
const PROJECT_CURRENT_FILE_NAME: &'static str = "current.yaml";

fn project_config_directory_path<P: AsRef<Path>>(home: P) -> PathBuf {
    home.as_ref().join(PROJECT_CONFIG_DIRECTORY)
}

#[allow(dead_code)]
fn create_project_config_directory<P: AsRef<Path>>(home: P) -> Result<()> {
    let project_config_directory = project_config_directory_path(home);
    match create_dir(project_config_directory) {
        Ok(_) => Ok(()),
        Err(e) => Err(Error::from(e)),
    }
}

#[allow(dead_code)]
fn project_current_file_path<P: AsRef<Path>>(home: P) -> PathBuf {
    project_config_directory_path(home).join(PROJECT_CURRENT_FILE_NAME)
}

#[allow(dead_code)]
fn load_project_current_file<P: AsRef<Path>>(home: P) -> Result<Projects> {
    let path = project_current_file_path(home);
    let file = File::open(path)?;
    let buf = BufReader::new(file);
    serde_yaml::from_reader(buf).map_err(|e| Error::from(e))
}

/// Create or overwrite project config file.
#[allow(dead_code)]
fn save_project_current_file<P: AsRef<Path>>(home: P, projects: Projects) -> Result<()> {
    let path = project_current_file_path(home);
    let file = OpenOptions::new()
        .write(true)
        .create(true)
        .truncate(true)
        .open(path)?;
    let buf = BufWriter::new(file);
    serde_yaml::to_writer(buf, &projects)?;
    Ok(())
}

#[derive(Serialize, Deserialize, Debug)]
struct Project {
    name: String,

    #[serde(skip_serializing)]
    current_env: Option<String>,

    #[serde(skip_serializing)]
    git_secret_repo: Option<String>,
}

impl Project {
    #[allow(dead_code)]
    fn new<S: AsRef<str>>(name: S) -> Project {
        Project {
            name: String::from(name.as_ref()),
            current_env: None,
            git_secret_repo: None,
        }
    }
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Projects {
    #[serde(rename = "projects")]
    all: Vec<Project>,
}

#[cfg(test)]
mod tests {
    use crate::project::current::{
        create_project_config_directory, load_project_current_file, project_current_file_path,
        save_project_current_file, Project, Projects,
    };
    use insta::assert_debug_snapshot;
    use serde_yaml;
    use std::collections::HashMap;
    use std::fs::read_to_string;
    use utils::asset::Assets;
    use utils::test::before;
    use walkdir::WalkDir;

    #[test]
    fn test_save_project_current_file() {
        let config = before(
            "test_save_project_current_file",
            Assets::All(HashMap::new()),
        );
        let projects = Projects {
            all: vec![Project::new("project_1")],
        };
        let r = create_project_config_directory(&config.tmp_dir);
        assert!(r.is_ok());
        let r = save_project_current_file(&config.tmp_dir, projects);
        assert!(r.is_ok());
        let r = read_to_string(project_current_file_path(&config.tmp_dir)).unwrap();
        assert_eq!(r, String::from("---\nprojects:\n  - name: project_1"));
        let read_create_projects_file = load_project_current_file(&config.tmp_dir);
        assert_debug_snapshot!(read_create_projects_file);
        let projects = Projects { all: vec![] };
        let r = save_project_current_file(&config.tmp_dir, projects);
        assert!(r.is_ok());
        let r = read_to_string(project_current_file_path(&config.tmp_dir)).unwrap();
        assert_eq!(r, String::from("---\nprojects: []"));
        let read_overwrite_projects_file = load_project_current_file(&config.tmp_dir);
        assert_debug_snapshot!(read_overwrite_projects_file);
    }

    #[test]
    fn test_load_project_current_file() {
        let mut assets: HashMap<&str, &str> = HashMap::new();
        assets.insert(
            ".d4d/current.yaml",
            r#"
projects: []
"#,
        );
        let config = before("test_load_project_current_file", Assets::All(assets));
        let projects = load_project_current_file(&config.tmp_dir);
        assert_debug_snapshot!(projects);
    }

    #[test]
    fn test_create_project_config_directory() {
        let config = before(
            "create_project_config_directory",
            Assets::All(HashMap::new()),
        );
        let r = create_project_config_directory(&config.tmp_dir);
        assert!(r.is_ok());
        let r: Vec<_> = WalkDir::new(&config.tmp_dir).into_iter().collect();
        assert_debug_snapshot!(r.iter().count());
        let r = create_project_config_directory(&config.tmp_dir);
        assert!(r.is_err());
    }

    #[test]
    fn test_serialize_projects() {
        let root = Projects { all: vec![] };
        let s = serde_yaml::to_string(&root).unwrap();
        assert_debug_snapshot!(&s);
    }
}
