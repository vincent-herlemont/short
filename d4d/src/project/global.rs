use serde::{Deserialize, Serialize};
use std::fs::{create_dir, File, OpenOptions};
use std::io::{BufReader, BufWriter};
use std::path::{Path, PathBuf};
use utils::error::Error;
use utils::result::Result;

#[allow(dead_code)]
const PROJECT_CONFIG_DIRECTORY: &'static str = ".d4d";
#[allow(dead_code)]
const PROJECT_CURRENT_FILE_NAME: &'static str = "projects.yaml";

fn global_directory_path<P: AsRef<Path>>(home: P) -> PathBuf {
    home.as_ref().join(PROJECT_CONFIG_DIRECTORY)
}

#[allow(dead_code)]
fn create_global_directory<P: AsRef<Path>>(home: P) -> Result<()> {
    let path = global_directory_path(home);
    match create_dir(path) {
        Ok(_) => Ok(()),
        Err(e) => Err(Error::from(e)),
    }
}

#[allow(dead_code)]
fn global_file_path<P: AsRef<Path>>(home: P) -> PathBuf {
    global_directory_path(home).join(PROJECT_CURRENT_FILE_NAME)
}

#[allow(dead_code)]
fn read_global_file<P: AsRef<Path>>(home: P) -> Result<GlobalProjects> {
    let path = global_file_path(home);
    let file = File::open(path)?;
    let buf = BufReader::new(file);
    serde_yaml::from_reader(buf).map_err(|e| Error::from(e))
}

/// Create or overwrite project config file.
#[allow(dead_code)]
fn save_global_file<P: AsRef<Path>>(home: P, projects: &GlobalProjects) -> Result<()> {
    let path = global_file_path(home);
    let file = OpenOptions::new()
        .write(true)
        .create(true)
        .truncate(true)
        .open(path)?;
    let buf = BufWriter::new(file);
    serde_yaml::to_writer(buf, &projects)?;
    Ok(())
}

#[derive(Serialize, Deserialize, Debug)]
struct GlobalProject {
    name: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    path: Option<PathBuf>,

    #[serde(skip_serializing_if = "Option::is_none")]
    current_env: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    git_secret_repo: Option<String>,
}

impl GlobalProject {
    #[allow(dead_code)]
    fn new<S: AsRef<str>>(name: S) -> GlobalProject {
        GlobalProject {
            name: String::from(name.as_ref()),
            path: None,
            current_env: None,
            git_secret_repo: None,
        }
    }
}

#[derive(Serialize, Deserialize, Debug)]
pub struct GlobalProjects {
    #[serde(skip)]
    home_dir: PathBuf,

    #[serde(rename = "projects")]
    all: Vec<GlobalProject>,
}

impl GlobalProjects {
    pub fn new<P: AsRef<Path>>(home_dir: P) -> Result<GlobalProjects> {
        let home_dir = home_dir.as_ref().to_path_buf();
        match read_global_file(&home_dir) {
            Ok(global_file) => Ok(global_file),
            Err(_) => {
                let global_directory = global_directory_path(&home_dir);
                if !global_directory.exists() {
                    match create_global_directory(&home_dir) {
                        Ok(_) => (),
                        Err(err) => {
                            return Err(Error::wrap(
                                format!(
                                    "fail to create configuration directory {}",
                                    global_directory.to_string_lossy()
                                ),
                                err,
                            ));
                        }
                    }
                }
                // TODO: match for create err only if file does not exist.
                let global_projects = GlobalProjects {
                    home_dir: home_dir.to_owned(),
                    all: vec![],
                };
                match save_global_file(&home_dir, &global_projects) {
                    Ok(_) => Ok(global_projects),
                    Err(err) => Err(Error::wrap(
                        format!(
                            "fail to create global file to {}",
                            home_dir.to_string_lossy()
                        ),
                        err,
                    )),
                }
            }
        }
    }

    pub fn add<N, P>(&mut self, name: N, path: P) -> Result<()>
    where
        N: AsRef<str>,
        P: AsRef<Path>,
    {
        let path = PathBuf::from(path.as_ref())
            .canonicalize()
            .map_err(|e| Error::wrap("fail to get absolute path of : {}", Error::from(e)))?;

        self.all.push(GlobalProject {
            name: String::from(name.as_ref()),
            path: Some(path),
            current_env: None,
            git_secret_repo: None,
        });

        if let Err(err) = save_global_file(&self.home_dir, self) {
            Err(Error::wrap(
                format!(
                    "fail to save global file to : {}",
                    self.home_dir.to_string_lossy(),
                ),
                err,
            ))
        } else {
            Ok(())
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::project::global::{
        create_global_directory, global_file_path, read_global_file, save_global_file,
        GlobalProject, GlobalProjects,
    };
    use insta::assert_debug_snapshot;
    use std::collections::HashMap;
    use std::fs::read_to_string;
    use std::path::PathBuf;
    use utils::asset::Assets;
    use utils::test::before;
    use walkdir::WalkDir;

    #[test]
    fn test_save_global_file() {
        let config = before("test_save_global_file", Assets::All(HashMap::new()));
        let projects = GlobalProjects {
            home_dir: PathBuf::new(),
            all: vec![GlobalProject::new("project_1")],
        };
        let r = create_global_directory(&config.tmp_dir);
        assert!(r.is_ok());
        let r = save_global_file(&config.tmp_dir, &projects);
        assert!(r.is_ok());
        let r = read_to_string(global_file_path(&config.tmp_dir)).unwrap();
        assert_eq!(r, String::from("---\nprojects:\n  - name: project_1"));

        // Overwrite
        let projects = GlobalProjects {
            home_dir: PathBuf::new(),
            all: vec![],
        };
        let r = save_global_file(&config.tmp_dir, &projects);
        assert!(r.is_ok());
        let r = read_to_string(global_file_path(&config.tmp_dir)).unwrap();
        assert_eq!(r, String::from("---\nprojects: []"));
    }

    #[test]
    fn test_empty_read_global_file() {
        let mut assets: HashMap<&str, &str> = HashMap::new();
        assets.insert(
            ".d4d/projects.yaml",
            r#"
projects: []
"#,
        );
        let config = before("test_empty_read_global_file", Assets::All(assets));
        let projects = read_global_file(&config.tmp_dir);
        assert_debug_snapshot!(projects);
    }

    #[test]
    fn test_read_global_file() {
        let mut assets: HashMap<&str, &str> = HashMap::new();
        assets.insert(
            ".d4d/projects.yaml",
            r#"
projects:
    - name: project_1
    - name: project_2
      current_env: dev
    - name: project_3
      path: /todo/plop/
      current_env: dev
    - name: project_4
      current_env: prod
      git_secret_repo: "git@privategit.com" 
"#,
        );
        let config = before("test_read_global_file", Assets::All(assets));
        let projects = read_global_file(&config.tmp_dir);
        assert_debug_snapshot!(projects);
    }

    #[test]
    fn test_create_global_directory() {
        let config = before("create_global_directory", Assets::All(HashMap::new()));
        let r = create_global_directory(&config.tmp_dir);
        assert!(r.is_ok());
        let r: Vec<_> = WalkDir::new(&config.tmp_dir).into_iter().collect();
        assert_eq!(r.iter().count(), 2);
        let r = create_global_directory(&config.tmp_dir);
        assert!(r.is_err());
    }
}
