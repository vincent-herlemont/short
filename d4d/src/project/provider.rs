use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};


use utils::error::Error;
use utils::result::Result;

#[derive(Debug, Serialize, Deserialize)]
#[serde(tag = "name")]
pub enum ProviderCfg {
    #[serde(rename = "aws")]
    ConfAws(AwsCfg),

    #[serde(rename = "unknown")]
    None,
}

impl ProviderCfg {
    pub fn aws(&self) -> Result<&AwsCfg> {
        if let ProviderCfg::ConfAws(conf_aws) = self {
            Ok(conf_aws)
        } else {
            Err(Error::from("aws provider is not defined"))
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct AwsCfg {
    region: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    template_path: Option<PathBuf>,
}

impl AwsCfg {
    pub fn new<N: AsRef<str>>(region: N) -> Self {
        Self {
            region: region.as_ref().to_string(),
            template_path: None,
        }
    }

    pub fn set_template_path<P: AsRef<Path>>(&mut self, template_path: P) -> &Self {
        self.template_path = Some(template_path.as_ref().to_path_buf());
        self
    }

    pub fn template_path(&self) -> Result<PathBuf> {
        self.template_path.clone().ok_or(Error::from(format!(
            "template_path missing", // TODO: set template name
        )))
    }
}
