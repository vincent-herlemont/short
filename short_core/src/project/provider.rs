use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};

use short_utils::error::Error;
use short_utils::result::Result;

#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
#[serde(tag = "name")]
pub enum ProviderCfg {
    #[serde(rename = "aws")]
    ConfAws(AwsCfg),

    #[serde(rename = "unknown")]
    None,
}

impl ProviderCfg {
    pub fn aws(&self) -> Result<&AwsCfg> {
        if let ProviderCfg::ConfAws(conf_aws) = self {
            Ok(conf_aws)
        } else {
            Err(Error::from("aws provider is not defined"))
        }
    }
}

#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
pub struct AwsCfg {
    #[serde(skip_serializing_if = "Option::is_none")]
    template_path: Option<PathBuf>,
}

impl AwsCfg {
    pub fn new() -> Self {
        Self {
            template_path: None,
        }
    }

    pub fn set_template_path<P: AsRef<Path>>(&mut self, template_path: P) -> &Self {
        self.template_path = Some(template_path.as_ref().to_path_buf());
        self
    }

    pub fn template_path(&self) -> Result<PathBuf> {
        self.template_path.clone().ok_or(Error::from(format!(
            "template_path missing", // TODO: set template name
        )))
    }
}
