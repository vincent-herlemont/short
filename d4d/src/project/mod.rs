use serde::{Deserialize, Serialize};
use std::fs::{create_dir, File};
use std::io::BufReader;
use std::path::{Path, PathBuf};
use utils::error::Error;
use utils::result::Result;

#[allow(dead_code)]
const PROJECT_CONFIG_DIRECTORY: &'static str = ".d4d";
#[allow(dead_code)]
const PROJECT_CONFIG_FILE_NAME: &'static str = "projects.yaml";

fn project_config_directory_path<P: AsRef<Path>>(home: P) -> PathBuf {
    home.as_ref().join(PROJECT_CONFIG_DIRECTORY)
}

#[allow(dead_code)]
fn create_project_config_directory<P: AsRef<Path>>(home: P) -> Result<()> {
    let project_config_directory = project_config_directory_path(home);
    match create_dir(project_config_directory) {
        Ok(_) => Ok(()),
        Err(e) => Err(Error::from(e)),
    }
}

#[allow(dead_code)]
fn project_config_file_path<P: AsRef<Path>>(home: P) -> PathBuf {
    project_config_directory_path(home).join(PROJECT_CONFIG_FILE_NAME)
}

#[allow(dead_code)]
fn load_project_config_file<P: AsRef<Path>>(home: P) -> Result<Projects> {
    let project_config = project_config_file_path(home);
    let file = File::open(project_config)?;
    let buf = BufReader::new(file);
    serde_yaml::from_reader(buf).map_err(|e| Error::from(e))
}

#[derive(Serialize, Deserialize, Debug)]
struct Project {
    name: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Projects {
    #[serde(rename = "projects")]
    all: Vec<Project>,
}

#[cfg(test)]
mod tests {
    use crate::project::{create_project_config_directory, load_project_config_file, Projects};
    use insta::assert_debug_snapshot;
    use serde_yaml;
    use std::collections::HashMap;
    use utils::test::before;
    use walkdir::WalkDir;

    #[test]
    fn test_load_project_config_file() {
        let mut assets: HashMap<&str, &str> = HashMap::new();
        assets.insert(
            "assets/.d4d/projects.yaml",
            r#"
projects: []
"#,
        );
        let config = before("test_load_project_config_directory", &assets);
        let projects = load_project_config_file(&config.tmp_dir);
        assert_debug_snapshot!(projects);
    }

    #[test]
    fn test_create_project_config_directory() {
        let config = before("create_project_config_directory", &HashMap::new());
        let r = create_project_config_directory(&config.tmp_dir);
        assert_debug_snapshot!(r);
        let r: Vec<_> = WalkDir::new(&config.tmp_dir).into_iter().collect();
        assert_debug_snapshot!(r.iter().count());
        let r = create_project_config_directory(&config.tmp_dir);
        assert_debug_snapshot!(r);
    }

    #[test]
    fn test_serialize_projects() {
        let root = Projects { all: vec![] };
        let s = serde_yaml::to_string(&root).unwrap();
        assert_debug_snapshot!(&s);
    }
}
