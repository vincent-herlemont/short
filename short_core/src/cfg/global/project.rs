use std::cell::RefCell;
use std::path::PathBuf;
use std::rc::Rc;

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};

use crate::cfg::global::setup::GlobalProjectSetupCfg;
use crate::cfg::{ProjectCfg, SetupCfg, SetupsCfg};

#[derive(Debug, Serialize, Deserialize)]
pub struct GlobalProjectCfg {
    file: PathBuf,
    setups: Rc<RefCell<Vec<Rc<RefCell<GlobalProjectSetupCfg>>>>>,
}

impl GlobalProjectCfg {
    pub fn new(file: &PathBuf) -> Result<Self> {
        let mut gp = GlobalProjectCfg {
            file: PathBuf::new(),
            setups: Rc::new(RefCell::new(vec![])),
        };
        gp.set_file(file)?;
        Ok(gp)
    }

    pub fn set_file(&mut self, file: &PathBuf) -> Result<()> {
        if (!file.is_absolute()) {
            return Err(anyhow!(format!(
                "project file path can not be relative {}",
                file.to_string_lossy()
            )));
        }
        if let None = file.file_name() {
            return Err(anyhow!(format!("project file has no name")));
        }
        self.file = file.clone();
        Ok(())
    }
}

impl ProjectCfg for GlobalProjectCfg {
    fn path(&self) -> PathBuf {
        self.file.to_owned()
    }
}

impl SetupsCfg for GlobalProjectCfg {
    type Setup = GlobalProjectSetupCfg;

    fn get_setups(&self) -> Rc<RefCell<Vec<Rc<RefCell<Self::Setup>>>>> {
        Rc::clone(&self.setups)
    }
}

#[cfg(test)]
mod test {
    use std::path::PathBuf;

    use crate::cfg::global::project::GlobalProjectCfg;
    use crate::cfg::global::setup::GlobalProjectSetupCfg;
    use crate::cfg::{EnvPathCfg, SetupsCfg};

    #[test]
    fn global_update_private_env_dir() {
        let setup_cfg = GlobalProjectSetupCfg::new("setup".into());

        let mut project_cfg = GlobalProjectCfg::new(&"/project".into()).unwrap();
        project_cfg.add_setup(setup_cfg);

        assert!(project_cfg.get_setups().borrow().iter().count().eq(&1));

        {
            let setup_cfg = project_cfg.get_setup("setup".into()).unwrap();
            setup_cfg
                .borrow_mut()
                .set_env_path_op(Some("/private_env".into()));
        }

        let global_project_setup_cfg_1 = project_cfg.get_setup("setup".into()).unwrap();
        assert_eq!(
            global_project_setup_cfg_1.borrow().env_path(),
            PathBuf::from("/private_env")
        );

        project_cfg.remove_by_name_setup("setup".into());
        assert!(project_cfg.get_setup("setup".into()).is_none());
    }
}
