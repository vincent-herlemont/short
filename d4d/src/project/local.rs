use crate::project::provider::{AwsCfg, ProviderCfg};
use serde::export::Formatter;
use serde::{Deserialize, Serialize};
use std::fmt;
use std::fmt::Display;
use std::fs::OpenOptions;
use std::io::{BufReader, BufWriter};
use std::path::{Path, PathBuf};

use utils::error::Error;

use utils::result::Result;

const PROJECT_FILE_NAME: &'static str = "d4d.yaml";

fn local_file_path<P: AsRef<Path>>(root: P) -> PathBuf {
    root.as_ref().join(PROJECT_FILE_NAME)
}

fn save_local_file<P: AsRef<Path>>(root: P, local_projects: &LocalProjects) -> Result<()> {
    let path = local_file_path(root);
    let file = OpenOptions::new()
        .write(true)
        .create(true)
        .truncate(true)
        .open(&path)?;
    let buf = BufWriter::new(file);
    serde_yaml::to_writer(buf, &local_projects).map_err(|err| {
        Error::wrap(
            format!("fail to save project file {}", path.to_string_lossy()),
            Error::from(err),
        )
    })
}

fn read_local_file<P: AsRef<Path>>(root: P) -> Result<LocalProjects> {
    let current_dir = PathBuf::from(root.as_ref());
    let local_file = local_file_path(&current_dir);
    let file = OpenOptions::new().read(true).open(&local_file)?;
    let buf = BufReader::new(file);
    serde_yaml::from_reader(buf)
        .map_err(|err| {
            Error::wrap(
                format!("fail to read project file {}", local_file.to_string_lossy()),
                Error::from(err),
            )
        })
        .map(|local_projects: LocalProjects| LocalProjects {
            current_dir,
            ..local_projects
        })
}

#[derive(Serialize, Deserialize, Debug)]
pub struct LocalProject {
    name: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    public_env_directory: Option<PathBuf>,

    provider: ProviderCfg,
}

impl LocalProject {
    pub fn name(&self) -> String {
        self.name.to_owned()
    }

    pub fn public_env_directory(&self) -> Option<PathBuf> {
        self.public_env_directory.to_owned()
    }

    pub fn provider(&self) -> &ProviderCfg {
        &self.provider
    }
}

impl Display for LocalProject {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.name)
    }
}

#[derive(Serialize, Deserialize, Debug)]
pub struct LocalProjects {
    #[serde(skip)]
    current_dir: PathBuf,

    #[serde(rename = "projects")]
    all: Vec<Box<LocalProject>>,
}

impl Display for LocalProjects {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        for local_project in &self.all {
            writeln!(f, " - {}", local_project)?;
        }
        Ok(())
    }
}

impl LocalProjects {
    pub fn new<P: AsRef<Path>>(current_dir: P) -> Result<LocalProjects> {
        let current_dir = current_dir.as_ref().to_path_buf();
        match read_local_file(&current_dir) {
            Ok(local_projects) => Ok(local_projects),
            Err(error) => {
                if error.is(|err| matches!(err, Error::SerdeYaml(_))) {
                    return Err(error);
                }
                // TODO: match for create err only if file does not exist.
                let local_projects = LocalProjects {
                    current_dir: current_dir.to_owned(),
                    all: vec![],
                };
                match save_local_file(&current_dir, &local_projects) {
                    Ok(_) => Ok(local_projects),
                    Err(err) => Err(Error::wrap(
                        format!(
                            "fail to create local file {}",
                            current_dir.to_string_lossy()
                        ),
                        err,
                    )),
                }
            }
        }
    }

    pub fn add<N, PED>(
        &mut self,
        name: N,
        public_env_directory: PED,
        provider: ProviderCfg,
    ) -> Result<()>
    where
        N: AsRef<str>,
        PED: AsRef<Path>,
    {
        self.all.push(Box::new(LocalProject {
            name: String::from(name.as_ref()),
            public_env_directory: Some(PathBuf::from(public_env_directory.as_ref())),
            provider,
        }));

        if let Err(err) = save_local_file(&self.current_dir, self) {
            Err(Error::wrap(
                format!(
                    "fail to save local file : {}",
                    self.current_dir.to_string_lossy()
                ),
                err,
            ))
        } else {
            Ok(())
        }
    }

    pub fn get<P: AsRef<str>>(&self, project_name: P) -> Option<&LocalProject> {
        self.all.iter().find_map(|local_project| {
            if local_project.name == project_name.as_ref() {
                Some(local_project.as_ref())
            } else {
                None
            }
        })
    }

    pub fn fake() -> Self {
        let aws_cfg = AwsCfg::new("us-east-1");
        Self {
            current_dir: PathBuf::from("/path/to/local"),
            all: vec![
                Box::new(LocalProject {
                    name: String::from("project_test"),
                    public_env_directory: None,
                    provider: ProviderCfg::ConfAws(
                        aws_cfg
                            .clone()
                            .set_template_path("./project_test.tpl")
                            .to_owned(),
                    ),
                }),
                Box::new(LocalProject {
                    name: String::from("project_test_bis"),
                    public_env_directory: None,
                    provider: ProviderCfg::ConfAws(
                        aws_cfg
                            .clone()
                            .set_template_path("./project_test_bis.tpl")
                            .to_owned(),
                    ),
                }),
            ],
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::project::local::{local_file_path, read_local_file, save_local_file, LocalProjects};
    use insta::assert_yaml_snapshot;
    use std::collections::HashMap;
    use std::fs::read_to_string;
    use std::path::PathBuf;
    use utils::asset::Assets;
    use utils::test::before;

    #[test]
    fn test_read_local_file() {
        let mut assets = HashMap::new();
        assets.insert(
            "d4d.yaml",
            r#"
projects:
    - name: test_1
      provider:
        name: aws
        region: us-east-3
    - name: test_2
      provider:
        name: aws
        region: us-east-3
        template_path: "./test_template.yaml"
        "#,
        );
        let config = before("test_save_local_file", Assets::Static(assets));
        let local_projects = read_local_file(&config.tmp_dir).unwrap();
        assert_eq!(&local_projects.current_dir, &config.tmp_dir);
        assert_yaml_snapshot!(local_projects);
    }

    #[test]
    fn test_save_local_file() {
        let config = before("test_save_local_file", Assets::Static(HashMap::new()));
        let local_projects = LocalProjects::fake();
        let r = save_local_file(&config.tmp_dir, &local_projects);
        assert!(r.is_ok());
        let content = read_to_string(local_file_path(&config.tmp_dir)).unwrap();
        assert_eq!(
            content,
            r#"---
projects:
  - name: project_test
    provider:
      name: aws
      region: us-east-1
      template_path: "./project_test.tpl"
  - name: project_test_bis
    provider:
      name: aws
      region: us-east-1
      template_path: "./project_test_bis.tpl""#
        );

        // Overwrite
        let local_projects = LocalProjects {
            current_dir: PathBuf::new(),
            all: vec![],
        };
        let r = save_local_file(&config.tmp_dir, &local_projects);
        assert!(r.is_ok());
        let content = read_to_string(local_file_path(&config.tmp_dir)).unwrap();
        assert_eq!(content, String::from("---\nprojects: []"));
    }
}
