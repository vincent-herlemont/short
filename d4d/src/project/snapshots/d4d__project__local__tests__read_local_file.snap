---
source: d4d/src/project/local.rs
expression: local_projects
---
projects:
  - name: test_1
    provider:
      name: aws
      region: us-east-3
  - name: test_2
    provider:
      name: aws
      region: us-east-3
      template_path: "./test_template.yaml"
